### PR TITLE
[MRG] Improving `MinHash.remove_many(...)` performance

### LINKED ISSUE #1553: Improving `remove_many` performance
The [current implementation][0] of `remove_many` takes a list of hashes (in Python) and copies it into Rust for removal. There are a few improvements to make it more performant, specifically allowing the use of `MinHash` sketches as argument, and call a different Rust function that avoids the list copying (similar to what [`add_many`][1] already does).

Tagging @keyabarve because she has been looking for a Rust issue for some time, and this will touch Python, Rust and the interface inbetween =]

Suggested plan for fixing this issue:

- Add tests for the new behavior (a `remove_many` taking a `MinHash` as argument). These tests will fail, because there is no implementation of `remove_many` that accepts `MinHash` sketches as arguments yet =])
- Once there is a test, modify the `remove_many` method to also take `MinHash` arguments. This will involve creating a Rust function and expose it to Python. (more details to be added here)
- Tests should be passing at this point.
- Add benchmarks for `remove_many` in [`benchmarks/benchmarks.py`][2]. Note there is a mem benchmark for [`add_many`][3] that can serve as template, but we want to check both `remove_many` with `list`s/`set`s or with a `MinHash` as arguments.
- Verify that new version of `remove_many` is actually faster than the previous version.

[0]: https://github.com/dib-lab/sourmash/blob/516dc53b65727e8a1379b8dc2971eda158aedecf/src/sourmash/minhash.py#L307L309
[1]: https://github.com/dib-lab/sourmash/blob/516dc53b65727e8a1379b8dc2971eda158aedecf/src/sourmash/minhash.py#L296L305
[2]: https://github.com/dib-lab/sourmash/blob/516dc53b65727e8a1379b8dc2971eda158aedecf/benchmarks/benchmarks.py
[3]: https://github.com/dib-lab/sourmash/blob/516dc53b65727e8a1379b8dc2971eda158aedecf/benchmarks/benchmarks.py#L104

## Other tips

- [This post](https://blog.luizirber.org/2020/01/10/sourmash-pr/) has many similarities with this issue (since it is a walkthorugh for #826, which modified `add_many`, our example for modifications in this issue)

---

From https://github.com/dib-lab/sourmash/issues/1552#issuecomment-847890311:

> The true fix is making `remove_many` more efficient, because it is a common use case. At the moment `remove_many` needs a list of hashes to remove (which is convenient, because we can pass a `list` or `set` with the hashes), but many times we are actually pulling the hashes from a `MinHash` without any modifications, and in this case it is better to pass the full `MinHash` into Rust and process all the data there (avoiding extracting hashes from the MinHash, then building a list, and so on).
> This block needs to be changed:
> 
> into something that checks if the argument is a `MinHash`, and if so calls another Rust function, like `add_many` does:
>

### PR DESCRIPTION
**Pull request by mr-eyes**
* This PR improve the performance of the [Python:MinHash.remove_many](https://github.com/dib-lab/sourmash/blob/516dc53b65727e8a1379b8dc2971eda158aedecf/src/sourmash/minhash.py#L307L309) - [Rust:kmerminhash_remove_many](https://github.com/dib-lab/sourmash/blob/08b8cabdd7839002171ae884ddd55a0bb07e7991/src/core/src/ffi/minhash.rs#L147-L163)
    * **Current behavior:** Copying a Python list of hashes to Rust, then apply removal.
    * **Expected new behavior:** Pass the MinHash object by reference, similar to [add_many](https://github.com/dib-lab/sourmash/blob/516dc53b65727e8a1379b8dc2971eda158aedecf/src/sourmash/minhash.py#L296L305) then perform the removal.
* This PR Fixes #1553, and it's related to Ref: #1552.

:warning: This PR will touch all the layers (Python, C header, Rust).

---
*First time contributing: [https://orcid.org/0000-0002-3419-4785](https://orcid.org/0000-0002-3419-4785)*

### CODE REVIEW

**Comment by mr-eyes** [comment]
I am having a problem running `make benchmark` or `tox -e asv` for the new written benchmark function but here are the results of profiling the new function.

### Benchmark results

#### Time Profiling

```
Timer unit: 1e-06 s

Total time: 3.96304 s
File: profile_remove_many.py
Function: old at line 15

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    15                                           @profile
    16                                           def old():
    17         1    3962976.0 3962976.0    100.0      mh1.remove_many(l)
    18         1         67.0     67.0      0.0      assert len(mh1) == 0

Total time: 5.7e-05 s
File: profile_remove_many.py
Function: new at line 21

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    21                                           @profile
    22                                           def new():
    23         1         50.0     50.0     87.7      mh2.remove_many(rm_mh)
    24         1          7.0      7.0     12.3      assert len(mh2) == 0
```

---

#### Memory profiling

```
Filename: profile_remove_many.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    15 3909.449 MiB 3909.449 MiB           1   @profile
    16                                         def old():
    17 3909.656 MiB    0.207 MiB           1       mh1.remove_many(l)
    18 3909.656 MiB    0.000 MiB           1       assert len(mh1) == 0


Filename: profile_remove_many.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    21 3909.656 MiB 3909.656 MiB           1   @profile
    22                                         def new():
    23 3909.656 MiB    0.000 MiB           1       mh2.remove_many(rm_mh)
    24 3909.656 MiB    0.000 MiB           1       assert len(mh2) == 0
```

---

### Reproduce the profiling

**profile_remove_many.py**
```python
from sourmash import MinHash

l = list(range(100000000))

mh1 = MinHash(500, 21, track_abundance=False)
mh1.add_many(l)

mh2 = MinHash(500, 21, track_abundance=False)
mh2.add_many(l)

rm_mh = MinHash(500, 21, track_abundance=False)
rm_mh.add_many(l)


@profile
def old():
    mh1.remove_many(l)
    assert len(mh1) == 0


@profile
def new():
    mh2.remove_many(rm_mh)
    assert len(mh2) == 0

old()
new()
```

**reproduce**

```bash
# time
kernprof -l profile_remove_many.py
python -m line_profiler profile_remove_many.py.lprof

# memory
python -m memory_profiler profile_remove_many.py
```

**Comment by mr-eyes** [comment]
Another simple profiling focusing on time 

```
add_many(iterable) took: 362.20860481262207 ms
add_many(MinHash) took: 0.09489059448242188 ms
```

---

```python
from sourmash import MinHash
from time import time

l = list(range(int(10000000)))

mh1 = MinHash(500, 21, track_abundance=False)
mh1.add_many(l)

mh2 = MinHash(500, 21, track_abundance=False)
mh2.add_many(l)

rm_mh = MinHash(500, 21, track_abundance=False)
rm_mh.add_many(l)

def old():
    mh1.remove_many(l)
    assert len(mh1) == 0


def new():
    mh2.remove_many(rm_mh)
    assert len(mh2) == 0

t = time()
old()
print(f"add_many(iterable) took: {1000 * (time() - t)} ms")

t = time()
new()
print(f"add_many(MinHash) took: {1000 * (time() - t)} ms")
```

**Comment by mr-eyes** [comment]
Would you please review it? @luizirber  @ctb

**Comment by ctb** [comment]
This looks good to me but we should wait for @luizirber if we can :)

**Comment by luizirber** [comment]
> This looks good to me but we should wait for @luizirber if we can :)

Good? GOOD? This is awesome! It's five orders of magnitude faster! =]
(will actually look at the code now, but really liked the benchmarking, Mo!)

**Review by luizirber — APPROVED** [review]
Code looks great too, thanks @mr-eyes!

**Comment by luizirber** [comment]
For fixing the two checks:
- https://github.com/dib-lab/sourmash/pull/1571/checks?check_run_id=2759014261#step:4:11 needs a `cargo fmt` and commit changes;
- https://github.com/dib-lab/sourmash/pull/1571/checks?check_run_id=2759014381#step:6:6 needs `cbindgen` to be rerun, mostly because of formatting of `include/sourmash.h`. I'll add the change as a suggestion, just a sec

**Comment by ctb** [comment]
sorry, updated from latest before I saw luiz's latest comments! will wait to merge.

File: include/sourmash.h [hunk 1]
```diff
@@ -227,6 +227,8 @@ uint32_t kmerminhash_num(const SourmashKmerMinHash *ptr);
 
 void kmerminhash_remove_hash(SourmashKmerMinHash *ptr, uint64_t h);
 
+void kmerminhash_remove_from(SourmashKmerMinHash *ptr, const SourmashKmerMinHash *other);
+
```
Comment (luizirber) [inline include/sourmash.h]: Functions in `include/sourmash.h` need to be in lexicographical order =]

```suggestion
void kmerminhash_remove_from(SourmashKmerMinHash *ptr, const SourmashKmerMinHash *other);

void kmerminhash_remove_hash(SourmashKmerMinHash *ptr, uint64_t h);

```

File: src/core/src/sketch/minhash.rs [hunk 2]
```diff
@@ -415,6 +415,14 @@ impl KmerMinHash {
         };
     }
 
+
```
Comment (luizirber) [inline src/core/src/sketch/minhash.rs]: Extra empty line here, remove it:
```suggestion
```

**Pull request merged**